### PR TITLE
PULL_REQUEST_TEMPLATE: init

### PR DESCRIPTION
**Pull request by jagajaga**
See https://github.com/blog/2111-issue-and-pull-request-templates
cc @domenkozar

### CODE REVIEW

File: PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+- [ ] Tested via `nix.useChroot`.
+- [ ] Compiled on all platforms written in `meta` section.
```
Comment (FRidh) [inline PULL_REQUEST_TEMPLATE.md]: Suggest using `build` instead of `compiled`.

File: PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+- [ ] Compiled on other platform(s): .
+- [ ] Tested compilation of all pkgs that depend on this change.
```
Comment (FRidh) [inline PULL_REQUEST_TEMPLATE.md]: Perhaps add  (e.g. with `nox-review`)

Comment (jagajaga) [inline PULL_REQUEST_TEMPLATE.md]: I suppose I should add nox-review to contributing guide.

File: PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (jagajaga) [inline PULL_REQUEST_TEMPLATE.md]: :+1:

File: PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+- [ ] Was executed in runtime.
+- [ ] Fits `CONTRIBUTING.md`.
+
+###### Extra
+Fixes # .
```
Comment (jagajaga) [inline PULL_REQUEST_TEMPLATE.md]: Note to myself: extra new line here.

**Comment by jgillich** [comment]
I'm not a huge fan of this in the current form. Not a lot of people test PRs on multiple platforms, and that's why you have CI anyway. I also don't want to build half of nixpkgs myself when I update a library, that's, again, what CI is for. Fixed issues belong in the commit messages.

What exactly is the goal of this template?

**Comment by jagajaga** [comment]
The goal is to make people more responsible and orderly. And it's a good cheet sheet about the things to notice. Also that helps nixos official commiters to minimize their work.

**Comment by zimbatm** [comment]
As a contributor, how do I compile on all the other platforms ?

I agree that people who submit a PR should have at least ran the software on a single platform.

File: PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (henrytill) [inline PULL_REQUEST_TEMPLATE.md]: In that case, it should be `built`.

File: PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+- [ ] Tested via `nix.useChroot`.
+- [ ] Compiled on all platforms written in `meta` section.
+- [ ] Compiled on other platform(s): .
+- [ ] Tested compilation of all pkgs that depend on this change.
+- [ ] Was executed in runtime.
```
Comment (henrytill) [inline PULL_REQUEST_TEMPLATE.md]: Perhaps something like:

`[ ] Tested resultant binary executables` 

or 

`[ ] Tested execution of binary products` 

would better fit the pattern here...

**Comment by colemickens** [comment]
Small suggestion, if this is merged... can it also link to a NixOS Wiki page that contains an explanation of these and examples of how to do them? I'm afraid even I don't know how to do some of them despite having PRs merged.

File: PULL_REQUEST_TEMPLATE.md (hunk 4 above)
Comment (edolstra) [inline PULL_REQUEST_TEMPLATE.md]: Or just `Tested package`.

**Comment by edolstra** [comment]
Could you put this in `.github` to avoid cluttering the root directory?

As to the content, it's not clear whether this is a checklist of steps that the contributor *must* do. If that's the case, it seems rather heavy to require people to test on all supported platforms. (I never do that.)

File: PULL_REQUEST_TEMPLATE.md (hunk 4 above)
Comment (vcunat) [inline PULL_REQUEST_TEMPLATE.md]: `Tested package` IMHO doesn't make it very clear it's supposed to be runtime testing.

Comment (jagajaga) [inline PULL_REQUEST_TEMPLATE.md]: I'm for `Tested execution of binary products`.

**Comment by jagajaga** [comment]
@edolstra sure will add to `.github`

Please note that this is not a list of all the things contributors should do. But this is a list of points that are good to help us.

**Comment by jagajaga** [comment]
Updated.

**Comment by jgillich** [comment]
Contributing should be easy, this makes it hard. I don't really see that many "bad" PRs that would make something like this necessary.

**Comment by jagajaga** [comment]
@jgillich that will *really* make our life easier. We need some organization of PRs.

**Comment by vcunat** [comment]
From the discussion I think we should make it explicit that the points aren't mandatory.

**Comment by jagajaga** [comment]
OK, will mention that.

-- 
Sincerely,
Arseniy Seroka

**Comment by jgillich** [comment]
@jagajaga You do review a lot of PRs, so if you say this would help, I'm going to believe you. From a contributor's perspective, I am concerned this could cause confusion and drive away contributors, but it's not for me to decide.

**Comment by zimbatm** [comment]
@jagajaga suggested changes: https://gist.github.com/zimbatm/16c4a56a3320c26e10f7

We can always try that out and refine the process as we go

**Comment by jgillich** [comment]
That looks better, although I'd omit the extra section. The "fixes" part should be in the commit message in order to auto close issues, with this you'd be referencing them twice.

**Comment by jagajaga** [comment]
@zimbatm unfortunately people won't read things. But if we try to force them put `x` near the point that will probably help. Example: there is also a note about `CONTRIBUTING.md` if you create a PR, but it seems that people just miss it..

**Comment by jagajaga** [comment]
Updated.

@jgillich that's just a template, so it will help people to write about references and mentions.
Yeah, also nobody cancels the references in a commit message :)

**Comment by FRidh** [comment]
Well, I wouldn't say they won't read. Sure, some don't. But there is also a lot of info out there that people simply will miss things. This checklist should be the absolute minimum that they should have checked.

Let's see how it works out. It's easy enough to change.

**Comment by jagajaga** [comment]
@FRidh sure! :)

**Comment by zimbatm** [comment]
I forgot to mention this: how do I test that all the depending packages also compile properly ?

Even if I read CONTRIBUTING.md, and then (half of people are dropped there) follow to https://hydra.nixos.org/build/31821726/download/1/nixpkgs/manual.html#chap-submitting-changes it doesn't tell me how to do this.

Also according to our manual each PR should be tested against master and staging. I don't think that in practice people are doing this are they ?

**Comment by FRidh** [comment]
@zimbatm you can use `nox-review` to test all dependents. It's indeed not mentioned. I never test against staging.

**Comment by jagajaga** [comment]
@zimbatm https://github.com/NixOS/nixpkgs/pull/13100#discussion_r53341222
Also its said in https://hydra.nixos.org/build/31821726/download/1/nixpkgs/manual.html#chap-submitting-changes that you can build your PR against your NixOS version, but its quiet obvious that you need to rabase against master if your PR is against it.

**Comment by zimbatm** [comment]
Okay. If nobody is testing against staging we should maybe change the manual too so that everything aligns.

**Comment by jagajaga** [comment]
@zimbatm `CONTRIBUTING.md` points to the right thing right now.

**Comment by jagajaga** [comment]
Going to merge and let's see the results. We can always improve the document anytime.

**Pull request merged**